scripts/commands/checkLinks.ts

```diff
@@ -0,0 +1,132 @@
+const path = require('node:path')
+const markdownLinkExtractor = require('markdown-link-extractor');
+
+const DOCS_ROOT = "./docs"
+const CONTENT_FILE_EXTENSIONS = [".md", ".mdx", ".ipynb"]
+
+class Link {
+  value: string
+  anchor: string
+  origin: string
+  isExternal: boolean
+
+  constructor(linkString: string, origin: string) {
+    /* 
+     * linkString: Link as it appears in source file
+     *     origin: Path to source file containing link
+     */
+
+    const splitLink = linkString.split('#', 1)
+    this.value = splitLink[0]
+    this.anchor = (splitLink.length > 1) ? `#${splitLink[1]}` : ''
+    this.origin = origin
+    this.isExternal = linkString.startsWith("http")
+  }
+
+  resolve(): string[] {
+    /*
+     * Return list of possible paths link could resolve to
+     */
+    if ( this.isExternal ) { return [ this.value ] }
+    if ( this.value === '' ) { return [ this.origin ] }
+    if ( this.value.startsWith("/images") ) {
+      return [ path.join("public/", this.value) ]
+    }
+
+    let baseFilePath
+    if (this.value.startsWith('/')) {
+      // Path is relative to DOCS_ROOT
+      baseFilePath = path.join(DOCS_ROOT, this.value)
+    } else {
+      // Path is relative to origin file
+      baseFilePath = path.join(
+        path.dirname(this.origin),
+        this.value
+      )
+    }
+    // Remove trailing '/' from path.join
+    baseFilePath = baseFilePath.replace(/\/$/gm, '')
+
+    // File may have one of many extensions (.md, .ipynb etc.), and/or be
+    // directory with an index file (e.g. `docs/build` should resolve to
+    // `docs/build/index.mdx`). We return a list of possible filenames.
+    let possibleFilePaths = []
+    for (let index of ['', '/index']) {
+      for (let extension of CONTENT_FILE_EXTENSIONS) {
+        possibleFilePaths.push(
+            baseFilePath + index + extension
+        )
+      }
+    }
+    return possibleFilePaths
+  }
+
+  check(): boolean {
+    /*
+     * True if link points to existing file, otherwise false
+     */
+    if (this.isExternal) {
+      // External link checking not supported yet
+      return true
+    }
+
+    for (let filePath of this.resolve()) {
+      if (existsSync(filePath)) {
+        return true
+      }
+    }
+
+    console.log(`❌ ${this.origin}: Could not find link '${this.value}'`)
+    return false
+  }
+}
+
+function markdownFromNotebook(source: string): string {
+  let markdown = ''
+  for (let cell of JSON.parse(source).cells) {
+    if (cell.source === 'markdown') {
+      markdown += cell.source
+    }
+  }
+  return markdown
+}
+
+function checkLinksInFile(filePath: string): boolean {
+  const source = readFileSync(filePath, {encoding: 'utf8'})
+  const markdown = (path.extname(filePath) === '.ipynb') ? markdownFromNotebook(source) : source
+  const links = markdownLinkExtractor(source).links.map((x: string) => new Link(x, filePath))
+
+  let allGood = true
+  for (let link of links) {
+    allGood = link.check() && allGood
+  }
```

Comment:
`return links.all(link => link.check())`

If you take the above suggestion to have `check()` return the error or `undefineed`, rather than directly logging it, then this would be `return links.map(link => link.check()).filter(Boolean)`, where `.filter(Boolean)` gets rid of the undefined.

Reply:
With `links.every`, the script will exit on the first broken link it finds, whereas the current loop prints _all_ broken links. Is exiting early ok?

Reply:
Ah, okay, never mind. We definitely do want to collect every single link. The `.map().filter()` suggestion would do that properly though, if you end up returning the error rather than logging directly.